### PR TITLE
add endpoint /_synapse/admin/v2/users/:user_id

### PR DESCRIPTION
**Pull request by jamtwister**
- GET query user information
- PUT create or modify user

### CODE REVIEW

File: src/v2.rs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+//! Endpoints in the `/_synapse/admin/v2/` scope.
+
+pub mod user_account;
```
Comment (jplatte) [inline src/v2.rs:3]: Please add a final newline in this file.

File: src/v2/user_account.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
+use std::time::SystemTime;
+
+use serde::{Serialize, Deserialize};
+
+use ruma::api::ruma_api;
+use ruma::identifiers::UserId;
+use ruma::api::client::r0::contact::get_contacts::ThirdPartyIdentifier;
+
+use ruma::api::client;
```
Comment (jplatte) [inline src/v2/user_account.rs:9]: Please merge the ruma imports into one and remove the newline between the serde and ruma imports (c.f. https://github.com/ruma/ruma/blob/main/CONTRIBUTING.md#import-formatting).

File: src/v2/user_account.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    fn default() -> Self {
+        UserInformation{
+            creation_ts: Some(SystemTime::now()),
+            ..Default::default()
+        }
+    }
+}
+
+
+pub mod query_user {
```
Comment (jplatte) [inline src/v2/user_account.rs:72]: This module should be in its own file, not an inline module declaration.

File: src/v2/user_account.rs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    impl Response {
+        /// Creates a new `Response` with all parameters defaulted.
+        pub fn new() -> Self {
+            Default::default()
+        }
+    }
+}
+
+
+pub mod create_modify_user {
```
Comment (jplatte) [inline src/v2/user_account.rs:119]: Same here.

File: src/v2/user_account.rs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        }
+
+        request: {
+            /// user ID for the account to renew
+            #[ruma_api(path)]
+            pub user_id: &'a UserId,
+
+            /// This is an optional parameter. Add this parameter to create an account or set this
+            /// password as new one for an existing account.
+            pub password: Option<String>,
```
Comment (jplatte) [inline src/v2/user_account.rs:140]:
```suggestion
            pub password: Option<&'a str>,
```

Same below.

File: src/v2/user_account.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+            // NOTE: Server explodes if attributes are not omitted but specified as null, like the default
+            // Serde case.
+
+            /// defaults to user_id, or the current value if user already exists
+            /// Some("") is treated as setting it to null.
+            #[serde(skip_serializing_if="Option::is_none")]
```
Comment (jplatte) [inline src/v2/user_account.rs:147]:
```suggestion
            #[serde(skip_serializing_if = "Option::is_none")]
```

here and below.

File: src/v2/user_account.rs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            /// Should the user be a server admin
+            /// defaults to false, or the current value if user already exists
+            #[serde(skip_serializing_if="Option::is_none")]
+            pub admin: Option<bool>,
+
+            /// Should the user be deactivated
+            /// defaults to false, or the current value if user already exists
+            #[serde(skip_serializing_if="Option::is_none")]
+            pub deactivated: Option<bool>,
+
```
Comment (jplatte) [inline src/v2/user_account.rs:167]:
```suggestion
```

File: src/v2/user_account.rs [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        }
+
+        #[derive(Default)]
```
Comment (jplatte) [inline src/v2/user_account.rs:170]: Since `UserInformation` doesn't implement `Default`, this won't work.

File: src/v2/user_account.rs [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        response: {
+            /// Information about the user.
+            #[ruma_api(body)]
+            pub info: UserInformation,
+        }
+
+        error: client::Error
```
Comment (jplatte) [inline src/v2/user_account.rs:177]: This is just a temporary workaround, right? If so, please add a comment stating that above it.

File: src/v2/user_account.rs [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+        // todo following to does are from synadminctl
+        // TODO: returns 200 if account-exist-and-was-updated,
+        // but 201 CREATED if a new account was created.
+        // However, ruma does throw away this information.
```
Comment (jplatte) [inline src/v2/user_account.rs:182]: Interesting. Yeah, there is no way in `ruma_api!` to distinguish these cases right now. I think unless this becomes a common pattern for Matrix, we should just implement the conversion of `http::Request` to `create_modify_user::Request` manually (https://github.com/ruma/ruma/issues/238 would help with that).

File: src/v2/user_account.rs (hunk 3 above)
Comment (jamtwister) [reply src/v2/user_account.rs:72]: ok, thought I put it into one file cause it is the same endpoint with GET & PUT.
I will separate them

**Comment by jplatte** [comment]
Given the low number of endpoint, I also think that we probably don't need any module nesting below `v1` / `v2`. The shared `UserInformation` type could just live in `v2`. WDYT?

File: src/v2/user_account.rs (hunk 8 above)
Comment (jamtwister) [reply src/v2/user_account.rs:170]: I implemented the default traif for this stuct, is this not sufficent?

Comment (jplatte) [reply src/v2/user_account.rs:170]: Oh, sorry! I didn't see a derive so thought there wouldn't be an implementation.

**Comment by jamtwister** [comment]
I am fine with that. Will move it

File: src/v2/user_account.rs (hunk 8 above)
Comment (jplatte) [reply src/v2/user_account.rs:170]: I just had a look. Your implementation unconditionally recurses into itself, and also given the type's nature I don't think there should be a `Default` impl. Generally that should only exist for types where there is an obvious default; this also goes for the request and response types in `ruma_api!` (so it should be used when there are no parameters or they're all optional / have default values).

File: src/v2/user_account.rs (hunk 9 above)
Comment (jamtwister) [reply src/v2/user_account.rs:177]: forgot to remove it after testing

File: src/v2/user_account.rs (hunk 8 above)
Comment (jamtwister) [reply src/v2/user_account.rs:170]: Ok, but how to implement the Response constructor then?

Comment (jplatte) [reply src/v2/user_account.rs:170]: By adding a `UserInformation` parameter to it.

File: src/v2/user_account.rs (hunk 3 above)
Comment (jamtwister) [reply src/v2/user_account.rs:69]: How should I continue here?
Regarding the discussion just before I think
`query_user_details::v2` and `create_modify_user::v2` ?
Beside this, should I put all the user admin endpoints into a submodule?

Comment (jplatte) [reply src/v2/user_account.rs:69]: I don't care much either way. You can keep putting things in `v2::` and we later sort it out, or you add new things with `v2` being the leaf module, and I'll transition the other ones over soon.

Re. grouping, I think a `users` module would be good, as that is part of the urls too and it seems pointless to repeat `admin` or `api` in the module name when that's part of the crate name already.

**Comment by jamtwister** [comment]
Like this? Create die module v2 inside the files.

**Comment by jplatte** [comment]
Could you please rebase this (`git rebase main` from your branch)? The diff shows a bunch of things that are already in the main branch, making this hard to review.

**Comment by jamtwister** [comment]
rebase done

**Comment by jplatte** [comment]
Okay, maybe your local `main` branch was outdated, since nothing really changed. Can you try `git rebase origin/main`?

**Comment by jamtwister** [comment]
main should have be up to date. done `git rebase origin/main`, but I think there are no changes. I cannot see changes which are already in main anymore???

File: src/users/common.rs [hunk 11]
```diff
@@ -0,0 +1,60 @@
+//! Common parts for the user endpoints.
```
Comment (jplatte) [inline src/users/common.rs:1]: Can you just move everything in this file to `src/users.rs`?

File: src/users/create_or_modify.rs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    api::{
+        client::{r0::contact::get_contacts::ThirdPartyIdentifier, Error},
+        ruma_api,
+    },
+    UserId,
+};
+
+pub mod v2 {
+    //! [GET /_synapse/admin/v2/users/:user_id](https://github.com/matrix-org/synapse/blob/master/docs/admin_api/user_admin_api.rst#create-or-modify-account)
+    use super::*;
```
Comment (jplatte) [inline src/users/create_or_modify.rs:14]: Please move `v2` into its own file and move the imports from above into it instead of using `super::*`.

File: src/users/get_details.rs [hunk 13]
```diff
@@ -0,0 +1,56 @@
+//! Endpoint to get details for a user account.
+
+pub use super::common::UserDetails;
+use ruma::{
+    api::{client::Error, ruma_api},
+    UserId,
+};
+
+pub mod v2 {
+    //! [GET /_synapse/admin/v2/users/:user_id](https://github.com/matrix-org/synapse/blob/master/docs/admin_api/user_admin_api.rst#query-user-account)
```
Comment (jplatte) [inline src/users/get_details.rs:10]: Same here.

File: src/v1/account_validity.rs [hunk 14]
```diff
@@ -1,8 +1,7 @@
 //! [GET /_synapse/admin/v1/account_validity/validity](https://github.com/matrix-org/synapse/blob/master/docs/admin_api/account_validity.rst)
 
-use std::time::SystemTime;
-
 use ruma::{api::ruma_api, UserId};
+use std::time::SystemTime;
```
Comment (jplatte) [inline src/v1/account_validity.rs:4]: Please revert this change.

File: src/users/get_details.rs [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+            /// Details about the user.
+            #[ruma_api(body)]
+            pub details: UserDetails,
+        }
+
+        // temporary workaround until
+        // https://github.com/matrix-org/matrix-rust-sdk/issues/125
+        // is solved
+        error: Error
+
```
Comment (jplatte) [inline src/users/get_details.rs:40]: Superfluous newline.

**Review by jplatte — COMMENTED** [review]
I don't understand what you did, it doesn't seem like a regular rebase since you didn't force-push (at least GitHub doesn't show any force-pushes), but it seems like the diff is no longer showing anything old.

Overall this looks good, I'd just like to request a few more changes.

**Comment by jamtwister** [comment]
I think I kown now what happend. Some how my different git identities and configs (of a totaly differen project) were mixed up.
 Within that config the IDE created a merge commit when I rebase from main.
